lib/HeroBonus.h

```diff
@@ -331,6 +331,8 @@ class DLL_LINKAGE CAddInfo : public std::vector<si32>
 	BONUS_NAME(SPECIAL_FIXED_VALUE_ENCHANT) /*specialty spell like Melody has, constant spell effect (i.e. 3 luck), additionalInfo = value to fix.*/\
 	BONUS_NAME(TOWN_MAGIC_WELL) /*one-time pseudo-bonus to implement Magic Well in the town*/\
 	BONUS_NAME(LIMITED_SHOOTING_RANGE) /*limits range of shooting creatures, doesn't adjust any other mechanics (half vs full damage etc). val - range in hexes, additional info - optional new range for broken arrow mechanic */\
+	BONUS_NAME(LEARN_ENEMY_SPELL) /*skill-agnostic eagle eye. subtype = 0 - chance, 1 - max spell level*/\

```

Comment:
Have no objections against this feature, but this form (subtype = 0 -> chance & subtype = 1 -> max level) is quite different from our typical usage of bonuses. Usually subtype implies different "filter", but not completely different interpretation.

Perhaps it is better to add two completely separate bonuses - LEARN_ENEMY_SPELL_CHANCE and LEARN_ENEMY_SPELL_LEVEL?
Or use some other field, e.g. additional info to store max level? Not sure which approach is better.

Reply:
@IvanSavenko AFAIK this logic is also used in SECONDARY_SKILL_PREMY. Two separate bonuses is also bad, because we should have a learn spell filter there somehow.  How then it will work if we used only LEARN_ENEMY_SPELL_CHANCE or LEARN_ENEMY_SPELL_LEVEL? If they should work only together, they should be one bonus. 

But can we have more data in bonuses instead of subtype?

Reply:
you can have more data as addinfo, also your bonus comment doesn't indicate that you use val

Reply:
>  How then it will work if we used only LEARN_ENEMY_SPELL_CHANCE or LEARN_ENEMY_SPELL_LEVEL? 

And how will your current version work if only one bonus subtype is defined, e.g. subtype=0 exists but subtype=1 does not? How is that different from my suggestion?

> But can we have more data in bonuses instead of subtype?

Additional info (addinfo)

Reply:
@IvanSavenko My current version is not working normally, as were prev version with SECONDARY_SKILL_PREMY and SECONDARY_SKILL_VAL2. I will try to use addinfo, but a bit later.

Reply:
Another option - use subtype for spell level.
Meaning Basic Eagle eye will give bonuses with subtype 1 & 2, Advanced will add subtype 3, Expert adds 4.

All artifacts that only affect chance have subtype 0.

Total chance to learns spell of level X = value of bonuses with subtypes 0 & X.

This will also act similar to some of our existing bonuses, where subtype = 0 signifies "any" filter, and non-zero subtype signifies specific domain, e.g. "only ranged attacks = 1", "only melee attacks = 2"

Reply:
@IvanSavenko about specials: we can use just selector SECONDARY_SKILL for secondary skill special bonuses. Can we add updater limited for bonus source? I think if it is not here, it will very useful feature to add.

Reply:
Selectors probably won't work.
Point of selector is to select only some active bonuses from bonuses graph.
But specialties should work on per-bonus level.

Assume that we have bonuses like this:
```
type    val    source    valType
DAMAGE  0.2    SKILL     BASE
DAMAGE  0.1    OBJECT    BASE
DAMAGE  0.5    ARTIFACT  PERCENT_TO_ALL
```
Correct formula without specialties would be (0.2+0.1) * (1+0.5) = 0.45

We can use selector with source = SKILL, apply specialty, then select non-skill bonuses and add it all together. But it:
1) This won't produce desired result - our selector won't pick bonus with PERCENT_TO_ALL value type and as result percentage bonus won't be applied correctly. We will end up with (0.2) + (0.1) * (1+0.5) = 0.35
2) This need to be done in every place where bonus value is accessed. Not realistic.

We *can* however, add updater to every bonus created by skill, e.g. inside `CSkill::addNewBonus`. However there are two potential issues that might break this approach. It is possible that I am wrong at it won't be a problem, but I am not familiar enough with bonus system to say that for sure.
1) Is it possible for updater attached to skill to be affected by hero specialty? Skills don't belong to hero, they are separate node in the bonus system. Does bonus system provides access to hero context in this scenario?
2) How should we get value of specialty? Another bonus? If yes - then there might be potential recursion where we need to compute bonus value in order to compute bonus value. If no - then how should such specialties be configured?

These two points need investigation (or some other approach needs to be found).

Reply:
Regarding two bonuses system - does eagle eye related artifacts give you eagle eye skill itself or just amplifies existing one? I think that two bonuses are ok. If only one present - two options: either we have minimum value or it just will not work at all until you get the other bonus as well and I think this second option is completely ok.

Reply:
I think that spell level should be primary skill required for eagle eye. If its value is 0 - we can not learn spells. and secondary one just shows probability and does nothing without the first one. Things which give eagle eye level must also give minimum probability base value. but if there is no probability bonus at all we can use some hardcoded minimum base value for it as an option or output warning to logs and again do nothing

Reply:
@nullkiller spell level should not be noted anywhere, I think. I think we should do something "is learnable" on each spell, and then check this property regard a filter of each skill like scholar and eagle eye. Filter can be by level, by school or by other parameters.

Reply:
@nullkiller eagle eye arts increase probability to learn spell, but don't give skill.

@rilian-la-te but spell level IS noted in game, in skill description. Implementing "isLearnable" will:
1) Force us to add (almost) identical structure to every spell, .e.g `"isLearnable" : { "eagleEye" : "basic" }`. High probability of typos or broken spell due to a missing/incorrect definition.
2) Will make description of a skill invalid. Since skill specifically says "allows to learn spells up to level X"

It might make sense to add basic `"isLearnable" : true/false`, to prevent spells like Titan's Thunder from being learnable (not usre how vcmi handles those right now), but I don't think that full-fledged filter is a good idea.

Reply:
@IvanSavenko yes, is noted. But how I can mod eagle eye to make it work only on earth spells (but with probability 100% and any level)? With level approach - there is no way to do it.

Reply:
Make subtype check for school, so:
`type = EAGLE_EYE, subtype = 0` = all schools (H3-style)
`type = EAGLE_EYE, subtype = 1` = some specific school (not sure if we have constants for magic schools)

And when computing Eagle Eye VCMI should use "subtype 0 or subtype X" selector where X is spell school

Reply:
@IvanSavenko yes, we have, but it starts from 0, 0 is AIR)

Reply:
> @IvanSavenko yes, we have, but it starts from 0, 0 is AIR)

SPELL_DAMAGE_REDUCTION bonus uses -1 for "all schools"
MAGIC_SCHOOL_SKILL uses bitmap for list of schools (so 15 for all schools)

Reply:
@IvanSavenko will look for it. For now I want to find a good balance in modernizing VCMI)

Reply:
Idea - make EAGLE_EYE symbiotic with magic schools a hero has: like having Earth Magic with Eagle Eye increase chance to catch earth magic spells and basic one.

Reply:
@IvanSavenko done eagle eye as 2 bonuses. But dislike it due to relation to one gameplay action.